### PR TITLE
fix N+1 problem for Scale Tool

### PR DESCRIPTION
**Pull request by MorriganR**
fix N+1 problem for Scale Tool

### CODE REVIEW

File: src/app/qgsmaptoolscalefeature.cpp [hunk 1]
```diff
@@ -364,12 +364,20 @@ void QgsMapToolScaleFeature::applyScaling( double scale )
...
-    vlayer->getFeatures( QgsFeatureRequest().setFilterFid( id ) ).nextFeature( feat );
+    if ( !feat.hasGeometry() )
+      continue;
+
     QgsGeometry geom = feat.geometry();
-    geom.transform( t );
+    if ( !( geom.transform( t ) == QgsGeometry::Success ) )
+      continue;
+
+    QgsFeatureId id = feat.id();
```
Comment (elpaso) [inline src/app/qgsmaptoolscalefeature.cpp:380]:
```suggestion
    const QgsFeatureId id = feat.id();
```

**Comment by elpaso** [comment]
Code looks good to me but what does this actually fix? Is there an issue you can reference?

**Comment by MorriganR** [comment]
> Code looks good to me but what does this actually fix? Is there an issue you can reference?

QGIS and network latency #42230

**Comment by elpaso** [comment]
@MorriganR can you elaborate? Is this an optimization because with this patch you don't fetch attributes?

In general, it is helpful to give some more context to reviewers so they don't have to guess: how would this relate with the PR title "fix N+1 problem for Scale Tool".

**Comment by MorriganR** [comment]
Yes, of course this can be called optimization. I am don't fetch attributes, but that is not the main point. I receive all the necessary geometries in one request to the database, I do not make a separate request to the database for each object. This is a typical problem for ORM and is often referred to as an "N+1 problem".

**Review by elpaso — APPROVED** [review]

**Comment by roya0045** [comment]
As I based this on the rotation tool, I'm curious to know if that tool, or any other tool might need similar changes.

**Comment by MorriganR** [comment]
> As I based this on the rotation tool, I'm curious to know if that tool, or any other tool might need similar changes.

Move & Rotate Tools are already fixed, but the Vertex Tool and Undo  are very slow if you are editing multiple objects at the same time and there is a network delay between the DB and QGIS.

**Comment by roya0045** [comment]
Good to hear, thanks!

**Comment by nyalldawson** [comment]
Great work @MorriganR  -- these optimisations are much appreciated!

**Pull request merged**